### ISSUE TITLE
Enhancements: It would be good to expose the detailed connection states to the user

### ISSUE THREAD

**Issue by fans3210** [awslabs/aws-mobile-appsync-sdk-ios#42]
It would be good to expose the detailed connection states to public so that user can choose to manually fetch the missing messages between the disconnect-to-connected time window. Eg: Connecting, Connected, Not Observable(might be 'connecting' too although might not be observable), Disconnected etc.

**Comment by rohandubal** [comment]
Hello @fans3210 

Thanks for the feedback. Can you elaborate a little more? 

For e.g. Do you want these states at a client level or at a subscription level?  Can you elaborate the use-cases which you are targeting?

Thanks
Rohan

**Comment by fans3210** [comment]
@rohandubal . Thanks. I prefer this to be on server side. Eg: 1. If I use lambda resolver, I prefer I can get all the subscribers' connection status in the lambda function's event field like event.subscriberStatus which is like a dictionary or json   2. If I'm using other types of resolvers beside lambda one, I can get the dictionary/json from the resolver's template code like: $context.subscriberStatus. That's only my personal opinion. I don't know where is the best place to store such information. But it should be on server side since I can choose whether or not to send push notification in lambda functions.

**Comment by MarioBajr** [comment]
Hi @fans3210, I have a branch fixing this, waiting #35 to be merged to create the PR

**Comment by fans3210** [comment]
@MarioBajr . Amazing. Tks

**Comment by muellerfr** [comment]
@rohandubal Looks like this PR was merged a while back. Can this issue be closed?

**Comment by MarioBajr** [comment]
Hi @muellerfr, 
I believe that you're referring to this PR #75 , and it hasn't been merged yet. In order to create the observable subscriptions status PR on top of the master branch I need the #28 to be accepted and merged.

**Comment by palpatim** [comment]
@fans3210 

Regarding your use case description:

> UseCase:
> For a chat app, normally I will unsubscribe all topics when the app entered background to reduce the battery and data cost. By exposing the detailed connection status to the developer, when app enters foreground again, I can manually handle retrieval of the chat messages I missed during the disconnection period.

AppSync currently suspends network activity when it enters the background, and resumes when the app resumes foreground. In addition, the recently-added [Delta Sync](https://aws-amplify.github.io/docs/ios/api#delta-sync) feature basically does just what you specify: when creating a new `sync` subscription, the system will query for only the changes since the last successful sync.

Does that meet your needs for this use case?

If so, I'd like to scope down this feature request to providing better notification of an initial connection. Currently, there's no way to know that a subscription has been established and will start receiving updates, only that you've either a) received a new subscription message; or b) disconnected.

I propose a simplified state like:

```swift
public enum SubscriptionWatcherStatus {
    case connecting
    case connected
    case disconnected
    case error(AWSAppSyncSubscriptionError)
 }
```

and an appropriate refactoring of `AWSAppSyncSubscriptionError` to be enum, rather than struct, based, similar to what you show in [your gist](https://gist.github.com/MarioBajr/d297cc0144a6e8fca4f6bd559946484b#file-observablesubscriptions-diff-L35)

Thoughts?

**Comment by fans3210** [comment]
@palpatim Sorry I don't know this issue is still opened. U guys did amazing jobs. The refactor is totally fine. Will reopen this issue when someone else requires such feature again. Thanks!

**Comment by shannon-hager-skookum** [comment]
@palpatim I know this issue is closed but are you still planning to add the `SubscriptionWatcherStatus` enum?

**Comment by palpatim** [comment]
@shannon-hager-skookum 

Yes, I think it makes sense to reopen this, scoped to the simplified state I mentioned in the comment above. The biggest question in my mind is how we actually accomplish the "connected" state, since right now there's no really great hook from the underlying subscription connection that the service has begun fulfilling it. In testing, I'm working around this by inspecting the list of topics available on a subscription--once it moves to a non-empty list, then I'm interpreting that to mean the service is delivering subscriptions. But I'm not sure that's reliable, and I *am* sure I'd rather have a service-blessed mechanism for detecting that state. :) So part of this feature request will be to investigate the "right" way of fulfilling the "connected" state.

**Comment by achager** [comment]
@palpatim Would you be able to elaborate on what is considered the "connected" state in this scenario? Is it more than the web socket itself being open?

**Comment by palpatim** [comment]
There is a delay between the time the socket is opened/connected to the server and the time the service begins delivering subscription updates to that socket. In our integration tests, we rely on either a hardcoded timeout or the aforementioned topic count to accommodate that.

**Comment by palpatim** [comment]
Clarification on the proposed statuses:

```swift
public enum SubscriptionWatcherStatus {
    /// The subscription is in process of connecting
    case connecting

    /// The subscription has connected and is receiving events from the service
    case connected

    /// The subscription has been disconnected because of a lifecycle event or manual disconnect request
    case disconnected

    /// The subscription is in an error state. The enum's associated value will provide more details, including recovery options if available.
    case error(AWSAppSyncSubscriptionError)
 }
```

**Comment by achager** [comment]
@palpatim Those proposed status look good and should cover our use cases. Thanks!

**Comment by palpatim** [comment]
Working on an implementation of this now.

**Comment by palpatim** [comment]
This should be coming later today, but in the meantime, the API changes are as follows:

```swift
// New typealias:
public typealias SubscriptionStatusChangeHandler = (AWSAppSyncSubscriptionWatcherStatus) -> Void

// New status enum

/// The status of a SubscriptionWatcher
public enum AWSAppSyncSubscriptionWatcherStatus {
    /// The subscription is in process of connecting
    case connecting

    /// The subscription has connected and is receiving events from the service
    case connected

    /// The subscription has been disconnected because of a lifecycle event or manual disconnect request
    case disconnected

    /// The subscription is in an error state. The enum's associated value will provide more details, including recovery options if available.
    case error(AWSAppSyncSubscriptionError)
}


// New `statusChangeHandler` option to the AppSyncClient `subscribe()` method:

public func subscribe<Subscription: GraphQLSubscription>(subscription: Subscription,
                                                         queue: DispatchQueue = DispatchQueue.main,
                                                         statusChangeHandler: SubscriptionStatusChangeHandler? = nil,
                                                         resultHandler: @escaping SubscriptionResultHandler<Subscription>) throws -> AWSAppSyncSubscriptionWatcher<Subscription>?
```

The statusChangeHandler will be invoked upon status changes to the watcher's underlying MQTT client. This is specifically useful for monitoring when the watcher moves to `.connected` status, as it indicates that the watcher client has received a SUBACK from the MQTT broker and is receiving subscription messages from the service. At that point, it would be safe to update the UI with appropriate messaging, or issue a "catch up" query (in the case of a reconnect operation, as will be done automatically by Delta Sync in the `sync` method)

At this point, the changes are expected to be wholly additive and so shouldn't break any existing implementations.

**Comment by palpatim** [comment]
Sorry for missing the update on this issue, but we believe this was fixed by #196, and refined by @rohandubal on subsequent commits. I'm closing this issue but please do open a new one if you see use cases that aren't covered by this status reporting.
